spynnaker/pyNN/models/neuron/abstract_population_vertex.py

```diff
@@ -828,204 +890,174 @@ def reset_to_first_timestep(self):
-
-    def get_ring_buffer_shifts(self, incoming_projections):
-        """ Get the shift of the ring buffers for transfer of values into the
-            input buffers for this model.
+        if abs(value) < 1.0:
+            return DataType.S1615.closest_representable_value(value)
+        return 1 / (
+            DataType.S1615.closest_representable_value_above(1 / value))
 
-        :param list(~spynnaker.pyNN.models.Projection) incoming_projections:
-            The projections to consider in the calculations
-        :rtype: list(int)
+    def __calculate_min_weights(self, incoming_projections):
+        """ Calculate the minimum weights required to best represent all the
+            possible weights coming into this vertex
+
+        :param list(~.Projection) incoming_projections: incoming proj to vertex
+
+        :return: list of minimum weights
+        :rtype: list(float)
         """
+        # Initialise to a maximum value
+        min_weights = [sys.maxsize for _ in range(
+            self.__neuron_impl.get_n_synapse_types())]
+
+        # Get the (global) weight_scale from the input_type in the neuron_impl
         weight_scale = self.__neuron_impl.get_global_weight_scale()
-        weight_scale_squared = weight_scale * weight_scale
-        # This only gets ring buffer shifts for neuron synapses
-        n_synapse_types = self.__neuron_impl.get_n_synapse_types()
-        running_totals = [RunningStats() for _ in range(n_synapse_types)]
-        delay_running_totals = [RunningStats() for _ in range(n_synapse_types)]
-        total_weights = numpy.zeros(n_synapse_types)
-        biggest_weight = numpy.zeros(n_synapse_types)
-        weights_signed = False
-        rate_stats = [RunningStats() for _ in range(n_synapse_types)]
-        steps_per_second = MICRO_TO_SECOND_CONVERSION / machine_time_step()
 
         for proj in incoming_projections:
             synapse_info = proj._synapse_information
             # Skip if this is a synapse dynamics synapse type
             if synapse_info.synapse_type_from_dynamics:
                 continue
             synapse_type = synapse_info.synapse_type
+            connector = synapse_info.connector
             synapse_dynamics = synapse_info.synapse_dynamics
+
+            weight_min = synapse_dynamics.get_weight_minimum(
+                connector, self.__weight_random_sigma, synapse_info)
+
+            if weight_min == 0:
+                weight_min = DataType.S1615.decode_from_int(1)
+            weight_min *= weight_scale
+            if not numpy.isnan(weight_min):
+                if min_weights[synapse_type] != sys.maxsize:
+                    weight_min = float_gcd(
+                        min_weights[synapse_type], weight_min)
+                min_weights[synapse_type] = min(
+                    min_weights[synapse_type], weight_min)
+
+            if isinstance(synapse_dynamics, SynapseDynamicsSTDP):
```

Comment:
It would be cleaner if this code could be pushed into the synapse_dynamics rather than doing different stuff here based on the isinstance.

Reply:
Code moved into relevant synapse dynamics.